packages/nuxt/src/core/app.ts

```diff
@@ -125,20 +125,31 @@ async function resolveApp (nuxt: Nuxt, app: NuxtApp) {
     }))
   }
 
-  // Resolve plugins
-  app.plugins = [
-    ...nuxt.options.plugins.map(normalizePlugin)
-  ]
+  // Re-initialise plugins
+  app.plugins = nuxt.options.plugins.map(normalizePlugin)
+
+  // Track scanned plugins separately so we can apply a sorting mechanism to them
+  const scannedPlugins: [string, NuxtPlugin][] = []
   for (const config of nuxt.options._layers.map(layer => layer.config)) {
-    app.plugins.push(...[
-      ...(config.plugins || []),
-      ...config.srcDir
-        ? await resolveFiles(config.srcDir, [
-          `${config.dir?.plugins || 'plugins'}/*.{ts,js,mjs,cjs,mts,cts}`,
-          `${config.dir?.plugins || 'plugins'}/*/index.*{ts,js,mjs,cjs,mts,cts}` // TODO: remove, only scan top-level plugins #18418
-        ])
-        : []
-    ].map(plugin => normalizePlugin(plugin as NuxtPlugin)))
+    for (const plugin of config.plugins || []) {
+      app.plugins.push(normalizePlugin(plugin as NuxtPlugin | string))
+    }
+
+    if (config.srcDir) {
+      const pluginDir = join(config.srcDir, config.dir?.plugins || 'plugins')
+      for (const file of await resolveFiles(pluginDir, [
+        '*.{ts,js,mjs,cjs,mts,cts}',
+        '*/index.*{ts,js,mjs,cjs,mts,cts}' // TODO: remove, only scan top-level plugins #18418
+      ])) {
+        scannedPlugins.push([file.replace(pluginDir, ''), normalizePlugin(file)])
+      }
+    }
+  }
+
+  // Sort scanned plugins by their names so a scanned `01.plugin.ts` will always run before finished `02.plugin.ts`
+  // and append to the plugin array
+  for (const [_name, plugin] of scannedPlugins.sort(([a], [b]) => a.localeCompare(b))) {
```

Comment:
uhmm, we should also apply order, you can have plugins there with object notation

Reply:
ok, order applied for client/server in `annotatePlugins`